klippy/extras/display/menu.py

```diff
@@ -0,0 +1,1137 @@
+
+    def get_gcode(self):
+        return self._get_formatted(self._gcode, self._input_value)
+
+    def is_editing(self):
+        return self._input_value is not None
+
+    def init_value(self):
+        self._input_value = None
+        if not self.is_readonly():
+            args = self._prepare_values()
+            if len(args) > 0:
+                try:
+                    self._input_value = float(args[0])
+                except:
+                    pass
+
+    def reset_value(self):
+        self._input_value = None
+
+    def inc_value(self):
+        if self._input_value is None:
+            return
+        if(self._reverse is True):
+            self._input_value -= abs(self._input_step)
+        else:
+            self._input_value += abs(self._input_step)
+        self._input_value = min(self._input_max, max(self._input_min, self._input_value))
+
+    def dec_value(self):
+        if self._input_value is None:
+            return
+        if(self._reverse is True):
+            self._input_value += abs(self._input_step)
+        else:
+            self._input_value -= abs(self._input_step)
+        self._input_value = min(self._input_max, max(self._input_min, self._input_value))
+
+class MenuGroup(MenuContainer):
+    def __init__(self, menu, config, sep = ','):
+        super(MenuGroup, self).__init__(menu, config)
+        self._sep = sep
+        self._show_back = False
+        self.selected = None
+        self.items = config.get('items')
+
+    def is_accepted(self, item):
+        return super(MenuGroup, self).is_accepted(item) and type(item) is not MenuCard
+
+    def is_scrollable(self):
+        return False
+
+    def is_enabled(self):
+        return not not len(self)
+
+    def is_readonly(self):
+        return all([item.is_readonly() for item in self._items])
+
+    def _names_aslist(self):
+        return self._aslist_split(self.items, sep=self._sep)
+
+    def init(self):
+        super(MenuGroup, self).init()
+        for item in self._items:
+            item.init()
+
+    def _render_item(self, item, selected = False, scroll = False):
+        name = "%s" % str(item.render(scroll))
+        if selected and not self.is_editing():
+            name = name if self._menu.blink_slow_state else ' '*len(name)
+        elif selected and self.is_editing():
+            name = name if self._menu.blink_fast_state else ' '*len(name)
+        return name
+
+    def _render(self):
+        s = ""
+        if self.selected is not None:
+            self.selected = (self.selected % len(self)) if len(self) > 0 else None
+
+        for i, item in enumerate(self):
+            s += self._render_item(item, (i == self.selected), True)
+        return s
+
+    def _call_selected(self, method = None):
+        res = None
+        try:
+            res = (getattr(self[self.selected], method)() if method is not None else self[self.selected])
+        except:
+            pass
```

Comment:
As a general rule, if an exception is raised, it should be logged via: `logging.exception("my msg")`